harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -425,43 +425,49 @@ def _compute_validation_metrics(self) -> workload.Response:
         metrics = {}  # type: Optional[Dict[str, Any]]
 
         if self._evaluate_batch_defined():
-            keys = None
-            batch_metrics = []
+            keys = None  # type: Optional[KeysView[str]]
+            reducers = {}
+            last_state = {}  # type: Dict[str, Any]
 
             self.validation_loader = cast(torch.utils.data.DataLoader, self.validation_loader)
             check.gt(len(self.validation_loader), 0)
             for batch in self.validation_loader:
                 batch = self.context.to_device(batch)
                 num_inputs += pytorch._data_length(batch)
 
-                vld_metrics = self.trial.evaluate_batch(batch=batch)
+                vld_metrics = self.trial.evaluate_batch(batch)
                 # Verify validation metric names are the same across batches.
                 if keys is None:
                     keys = vld_metrics.keys()
+                    reducers = self._prepare_metrics_reducers(keys)
                 else:
                     check.eq(
                         keys,
                         vld_metrics.keys(),
                         "Validation metric names must match across all batches of data.",
                     )
-                check.is_instance(
-                    vld_metrics,
-                    dict,
-                    "validation_metrics() must return a "
-                    "dictionary of string names to Tensor "
-                    "metrics",
-                )
                 # TODO: For performance perform -> cpu() only at the end of validation.
```

Comment:
non-blocking: this is very confusing. Can you remove that?

Reply:
See above about the XXX comment.  This will be removed one way or another, but it needs to be addressed first.